File: clients/go/mongocore/ops/ops.go
package ops

import (
	"fmt"

	pb "github.com/rozza/mongocore/clients/go/proto"
	"go.mongodb.org/mongo-driver/v2/bson"
)

// Helper function to encode BSON documents
func encodeBson(doc interface{}) ([]byte, error) {
	return bson.Marshal(doc)
}

// Find creates a Find operation for use in a pipeline.
func Find(database, collection string, filter bson.D, options *pb.FindOptions) (*pb.PipelineOperation, error) {
	filterBytes, err := encodeBson(filter)
	if err != nil {
		return nil, fmt.Errorf("failed to encode filter: %w", err)
	}

	return &pb.PipelineOperation{
		Operation: &pb.PipelineOperation_Find{
			Find: &pb.FindRequest{
				Database:   database,
				Collection: collection,
				Filter:     &pb.Filter{Data: filterBytes},
				Options:    options,
			},
		},
	}, nil
}

// FindOne creates a FindOne operation for use in a pipeline.
func FindOne(database, collection string, filter bson.D) (*pb.PipelineOperation, error) {
	filterBytes, err := encodeBson(filter)
	if err != nil {
		return nil, fmt.Errorf("failed to encode filter: %w", err)
	}

	return &pb.PipelineOperation{
		Operation: &pb.PipelineOperation_FindOne{
			FindOne: &pb.FindOneRequest{
				Database:   database,
				Collection: collection,
				Filter:     &pb.Filter{Data: filterBytes},
			},
		},
	}, nil
}

// Insert creates an Insert operation for use in a pipeline.
func Insert(database, collection string, document bson.D) (*pb.PipelineOperation, error) {
	docBytes, err := encodeBson(document)
	if err != nil {
		return nil, fmt.Errorf("failed to encode document: %w", err)
	}

	return &pb.PipelineOperation{
		Operation: &pb.PipelineOperation_Insert{
			Insert: &pb.InsertRequest{
				Database:   database,
				Collection: collection,
				Document:   &pb.Document{Data: docBytes},
			},
		},
	}, nil
}

// InsertMany creates an InsertMany operation for use in a pipeline.
func InsertMany(database, collection string, documents []bson.D) (*pb.PipelineOperation, error) {
	pbDocs := make([]*pb.Document, 0, len(documents))
	for _, doc := range documents {
		docBytes, err := encodeBson(doc)
		if err != nil {
			return nil, fmt.Errorf("failed to encode document: %w", err)
		}
		pbDocs = append(pbDocs, &pb.Document{Data: docBytes})
	}

	return &pb.PipelineOperation{
		Operation: &pb.PipelineOperation_InsertMany{
			InsertMany: &pb.InsertManyRequest{
				Database:   database,
				Collection: collection,
				Documents:  pbDocs,
			},
		},
	}, nil
}

// Update creates an Update operation for use in a pipeline.
func Update(database, collection string, filter, update bson.D, upsert bool) (*pb.PipelineOperation, error) {
	filterBytes, err := encodeBson(filter)
	if err != nil {
		return nil, fmt.Errorf("failed to encode filter: %w", err)
	}
	updateBytes, err := encodeBson(update)
	if err != nil {
		return nil, fmt.Errorf("failed to encode update: %w", err)
	}

	return &pb.PipelineOperation{
		Operation: &pb.PipelineOperation_Update{
			Update: &pb.UpdateRequest{
				Database:   database,
				Collection: collection,
				Filter:     &pb.Filter{Data: filterBytes},
				Update:     &pb.Document{Data: updateBytes},
				Upsert:     upsert,
			},
		},
	}, nil
}

// UpdateMany creates an UpdateMany operation for use in a pipeline.
func UpdateMany(database, collection string, filter, update bson.D, upsert bool) (*pb.PipelineOperation, error) {
	filterBytes, err := encodeBson(filter)
	if err != nil {
		return nil, fmt.Errorf("failed to encode filter: %w", err)
	}
	updateBytes, err := encodeBson(update)
	if err != nil {
		return nil, fmt.Errorf("failed to encode update: %w", err)
	}

	return &pb.PipelineOperation{
		Operation: &pb.PipelineOperation_UpdateMany{
			UpdateMany: &pb.UpdateManyRequest{
				Database:   database,
				Collection: collection,
				Filter:     &pb.Filter{Data: filterBytes},
				Update:     &pb.Document{Data: updateBytes},
				Upsert:     upsert,
			},
		},
	}, nil
}

// Delete creates a Delete operation for use in a pipeline.
func Delete(database, collection string, filter bson.D) (*pb.PipelineOperation, error) {
	filterBytes, err := encodeBson(filter)
	if err != nil {
		return nil, fmt.Errorf("failed to encode filter: %w", err)
	}

	return &pb.PipelineOperation{
		Operation: &pb.PipelineOperation_Delete{
			Delete: &pb.DeleteRequest{
				Database:   database,
				Collection: collection,
				Filter:     &pb.Filter{Data: filterBytes},
			},
		},
	}, nil
}

// DeleteMany creates a DeleteMany operation for use in a pipeline.
func DeleteMany(database, collection string, filter bson.D) (*pb.PipelineOperation, error) {
	filterBytes, err := encodeBson(filter)
	if err != nil {
		return nil, fmt.Errorf("failed to encode filter: %w", err)
	}

	return &pb.PipelineOperation{
		Operation: &pb.PipelineOperation_DeleteMany{
			DeleteMany: &pb.DeleteManyRequest{
				Database:   database,
				Collection: collection,
				Filter:     &pb.Filter{Data: filterBytes},
			},
		},
	}, nil
}

// Aggregate creates an Aggregate operation for use in a pipeline.
func Aggregate(database, collection string, pipeline []bson.D) (*pb.PipelineOperation, error) {
	stages := make([][]byte, 0, len(pipeline))
	for _, stage := range pipeline {
		stageBytes, err := encodeBson(stage)
		if err != nil {
			return nil, fmt.Errorf("failed to encode pipeline stage: %w", err)
		}
		stages = append(stages, stageBytes)
	}

	return &pb.PipelineOperation{
		Operation: &pb.PipelineOperation_Aggregate{
			Aggregate: &pb.AggregateRequest{
				Database:   database,
				Collection: collection,
				Pipeline:   &pb.Pipeline{Stages: stages},
			},
		},
	}, nil
}

// RunCommand creates a RunCommand operation for use in a pipeline.
func RunCommand(database string, command bson.D, allowAll bool) (*pb.PipelineOperation, error) {
	commandBytes, err := encodeBson(command)
	if err != nil {
		return nil, fmt.Errorf("failed to encode command: %w", err)
	}

	return &pb.PipelineOperation{
		Operation: &pb.PipelineOperation_RunCommand{
			RunCommand: &pb.RunCommandRequest{
				Database: database,
				Command:  &pb.Document{Data: commandBytes},
				AllowAll: allowAll,
			},
		},
	}, nil
}

// ListDatabases creates a ListDatabases operation for use in a pipeline.
func ListDatabases() *pb.PipelineOperation {
	return &pb.PipelineOperation{
		Operation: &pb.PipelineOperation_ListDatabases{
			ListDatabases: &pb.ListDatabasesRequest{},
		},
	}
}

// ListCollections creates a ListCollections operation for use in a pipeline.
func ListCollections(database string) *pb.PipelineOperation {
	return &pb.PipelineOperation{
		Operation: &pb.PipelineOperation_ListCollections{
			ListCollections: &pb.ListCollectionsRequest{
				Database: database,
			},
		},
	}
}

File: clients/java/src/main/java/com/mongocore/Ops.java
package com.mongocore;

import mongocore.v1.Mongocore;
import mongocore.v1.Types;
import org.bson.Document;
import com.google.protobuf.ByteString;
import org.bson.codecs.DocumentCodec;
import org.bson.codecs.EncoderContext;
import org.bson.io.BasicOutputBuffer;

import java.util.List;
import java.util.stream.Collectors;

/**
 * Operation builders for pipeline requests.
 * Each method returns a PipelineOperation proto message.
 */
public class Ops {
    private static final DocumentCodec CODEC = new DocumentCodec();

    private static ByteString encodeBson(Document doc) {
        BasicOutputBuffer buffer = new BasicOutputBuffer();
        CODEC.encode(new org.bson.BsonBinaryWriter(buffer), doc, EncoderContext.builder().build());
        return ByteString.copyFrom(buffer.getInternalBuffer(), 0, buffer.getSize());
    }

    private static Types.Filter makeFilter(Document filter) {
        return Types.Filter.newBuilder().setData(encodeBson(filter)).build();
    }

    private static Types.Document makeDocument(Document doc) {
        return Types.Document.newBuilder().setData(encodeBson(doc)).build();
    }

    public static Mongocore.PipelineOperation find(String database, String collection, Document filter) {
        return Mongocore.PipelineOperation.newBuilder()
                .setFind(Mongocore.FindRequest.newBuilder()
                        .setDatabase(database)
                        .setCollection(collection)
                        .setFilter(makeFilter(filter))
                        .build())
                .build();
    }

    public static Mongocore.PipelineOperation findOne(String database, String collection, Document filter) {
        return Mongocore.PipelineOperation.newBuilder()
                .setFindOne(Mongocore.FindOneRequest.newBuilder()
                        .setDatabase(database)
                        .setCollection(collection)
                        .setFilter(makeFilter(filter))
                        .build())
                .build();
    }

    public static Mongocore.PipelineOperation insert(String database, String collection, Document document) {
        return Mongocore.PipelineOperation.newBuilder()
                .setInsert(Mongocore.InsertRequest.newBuilder()
                        .setDatabase(database)
                        .setCollection(collection)
                        .setDocument(makeDocument(document))
                        .build())
                .build();
    }

    public static Mongocore.PipelineOperation insertMany(String database, String collection, List<Document> documents) {
        List<Types.Document> pbDocs = documents.stream()
                .map(Ops::makeDocument)
                .collect(Collectors.toList());

        return Mongocore.PipelineOperation.newBuilder()
                .setInsertMany(Mongocore.InsertManyRequest.newBuilder()
                        .setDatabase(database)
                        .setCollection(collection)
                        .addAllDocuments(pbDocs)
                        .build())
                .build();
    }

    public static Mongocore.PipelineOperation update(String database, String collection, Document filter, Document update) {
        return Mongocore.PipelineOperation.newBuilder()
                .setUpdate(Mongocore.UpdateRequest.newBuilder()
                        .setDatabase(database)
                        .setCollection(collection)
                        .setFilter(makeFilter(filter))
                        .setUpdate(makeDocument(update))
                        .build())
                .build();
    }

    public static Mongocore.PipelineOperation updateMany(String database, String collection, Document filter, Document update) {
        return Mongocore.PipelineOperation.newBuilder()
                .setUpdateMany(Mongocore.UpdateManyRequest.newBuilder()
                        .setDatabase(database)
                        .setCollection(collection)
                        .setFilter(makeFilter(filter))
                        .setUpdate(makeDocument(update))
                        .build())
                .build();
    }

    public static Mongocore.PipelineOperation delete(String database, String collection, Document filter) {
        return Mongocore.PipelineOperation.newBuilder()
                .setDelete(Mongocore.DeleteRequest.newBuilder()
                        .setDatabase(database)
                        .setCollection(collection)
                        .setFilter(makeFilter(filter))
                        .build())
                .build();
    }

    public static Mongocore.PipelineOperation deleteMany(String database, String collection, Document filter) {
        return Mongocore.PipelineOperation.newBuilder()
                .setDeleteMany(Mongocore.DeleteManyRequest.newBuilder()
                        .setDatabase(database)
                        .setCollection(collection)
                        .setFilter(makeFilter(filter))
                        .build())
                .build();
    }

    public static Mongocore.PipelineOperation aggregate(String database, String collection, List<Document> pipeline) {
        List<ByteString> stages = pipeline.stream()
                .map(Ops::encodeBson)
                .collect(Collectors.toList());

        return Mongocore.PipelineOperation.newBuilder()
                .setAggregate(Mongocore.AggregateRequest.newBuilder()
                        .setDatabase(database)
                        .setCollection(collection)
                        .setPipeline(Types.Pipeline.newBuilder().addAllStages(stages).build())
                        .build())
                .build();
    }

    public static Mongocore.PipelineOperation runCommand(String database, Document command, boolean allowAll) {
        return Mongocore.PipelineOperation.newBuilder()
                .setRunCommand(Mongocore.RunCommandRequest.newBuilder()
                        .setDatabase(database)
                        .setCommand(makeDocument(command))
                        .setAllowAll(allowAll)
                        .build())
                .build();
    }

    public static Mongocore.PipelineOperation listDatabases() {
        return Mongocore.PipelineOperation.newBuilder()
                .setListDatabases(Mongocore.ListDatabasesRequest.newBuilder().build())
                .build();
    }

    public static Mongocore.PipelineOperation listCollections(String database) {
        return Mongocore.PipelineOperation.newBuilder()
                .setListCollections(Mongocore.ListCollectionsRequest.newBuilder()
                        .setDatabase(database)
                        .build())
                .build();
    }
}
